numpy/core/src/multiarray/array_coercion.c

```diff
@@ -0,0 +1,1434 @@
+        coercion_cache_obj *current = *coercion_cache_head;  /* item to check */
+        while (current != NULL) {
+            if (current->depth > ndim) {
+                /* delete "next" cache item and advanced it (unlike later) */
+                current = npy_unlink_coercion_cache(current);
+                continue;
+            }
+            /* advance both prev and next, and set prev->next to new item */
+            *next_ptr = current;
+            next_ptr = &(current->next);
+            current = current->next;
+        }
+        *next_ptr = NULL;
+    }
+    /* We could check here for max-ndims being reached as well */
+
+    if (requested_descr != NULL) {
+        /* descriptor was provided, we did not accidentally change it */
+        assert(*out_descr == requested_descr);
+    }
+    else if (NPY_UNLIKELY(*out_descr == NULL)) {
+        /*
+         * When the object contained no elements (sequence of length zero),
+         * the no descriptor may have been found. When a DType was requested
+         * we use it to define the output dtype.
+         * Otherwise, out_descr will remain NULL and the caller has to set
+         * the correct default.
+         */
+        if (fixed_DType != NULL) {
+            if (fixed_DType->default_descr == NULL) {
+                Py_INCREF(fixed_DType->singleton);
+                *out_descr = fixed_DType->singleton;
+            }
+            else {
+                *out_descr = fixed_DType->default_descr(fixed_DType);
+                if (*out_descr == NULL) {
+                    goto fail;
+                }
+            }
+        }
+    }
+    return ndim;
+
+  fail:
+    npy_free_coercion_cache(*coercion_cache_head);
+    *coercion_cache_head = NULL;
+    Py_XSETREF(*out_descr, NULL);
+    return -1;
+}
+
+
+
+/**
+ * Check the descriptor is a legacy "flexible" DType instance, this is
+ * an instance which is (normally) not attached to an array, such as a string
+ * of length 0 or a datetime with no unit.
+ * These should be largely deprecated, and represent only the DType class
+ * for most `dtype` parameters.
+ *
+ * TODO: This function should eventually recieve a deprecation warning and
+ *       be removed.
+ *
+ * @param descr
+ * @return 1 if this is not a concrete dtype instance 0 otherwise
+ */
+static int
+descr_is_legacy_parametric_instance(PyArray_Descr *descr)
+{
+    if (PyDataType_ISUNSIZED(descr)) {
+        return 1;
+    }
+    /* Flexible descr with generic time unit (which can be adapted) */
+    if (PyDataType_ISDATETIME(descr)) {
+        PyArray_DatetimeMetaData *meta;
+        meta = get_datetime_metadata_from_dtype(descr);
+        if (meta->base == NPY_FR_GENERIC) {
+            return 1;
+        }
+    }
+    return 0;
+}
+
+
+/**
+ * Given either a DType instance or class, (or legacy flexible instance),
+ * ands sets output dtype instance and DType class. Both results may be
+ * NULL, but if `out_descr` is set `out_DType` will always be the
```

Comment:
The one branch that sets it, first sets `out_DType`?  The up-front NULL'ing is probably unnecessary (would have to put an `else` instead I guess.

Reply:
> The one branch that sets it, first sets `out_DType`? The up-front NULL'ing is probably unnecessary (would have to put an `else` instead I guess.

I  was assuming that if ``out_descr`` is set by caller then the out_DType should be accordingly set, but maybe there is no need for that use case to be supported here ?

Reply:
Oh, the user is only supposed to pass in `dtype` here. I guess its different from the discovery function, where it is possible to pass in a "starting" dtype because we use it in a few places.  Maybe a few words in the `@param`'s would help?